src/internal/task/etcd_task_service_test.go

```diff
@@ -0,0 +1,175 @@
+package task
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/require"
+	"github.com/pachyderm/pachyderm/v2/src/internal/testetcd"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	errTaskFailure = errors.Errorf("task failure")
+)
+
+func seedStr(seed int64) string {
+	return fmt.Sprint("seed: ", strconv.FormatInt(seed, 10))
+}
+
+func serializeTestTask(testTask *TestTask) (*types.Any, error) {
+	serializedTestTask, err := proto.Marshal(testTask)
+	if err != nil {
+		return nil, err
+	}
+	return &types.Any{
+		TypeUrl: "/" + string(proto.MessageName(testTask)),
+		Value:   serializedTestTask,
+	}, nil
+}
+
+func deserializeTestTask(any *types.Any) (*TestTask, error) {
+	testTask := &TestTask{}
+	if err := types.UnmarshalAny(any, testTask); err != nil {
+		return nil, err
+	}
+	return testTask, nil
+}
+
+func test(t *testing.T, workerFailProb, groupCancelProb, taskFailProb float64) {
+	seed := time.Now().UTC().UnixNano()
+	rand.Seed(seed)
+	msg := seedStr(seed)
+	env := testetcd.NewEnv(t)
+	numGroups := 10
+	numTasks := 10
+	numWorkers := 5
+	// Set up task service.
+	s := NewEtcdService(env.EtcdClient, "")
```

Comment:
You might consider adding an argument to test: `s Service`, and then creating a fresh etcd from `testetcd` as part of the setup of each test below.